path_checkouts.go

```diff
@@ -0,0 +1,390 @@
+		resp.Secret.TTL = ttl
+		resp.Secret.MaxTTL = set.MaxTTL
+		return resp, nil
+	}
+
+	// If we arrived here, it's because we never had a hit for a service account that was available.
+	// In case of customer issues, we need to make this easy to see and diagnose.
+	b.Logger().Debug(fmt.Sprintf(`%q had no check-outs available`, setName))
+	metrics.IncrCounter([]string{"ldap", "check-out", "unavailable", setName}, 1)
+	return logical.ErrorResponse("No service accounts available for check-out."), nil
+}
+
+func checkoutSecretCreds(b *backend) *framework.Secret {
+	return &framework.Secret{
+		Type: checkoutKeyType,
+		Fields: map[string]*framework.FieldSchema{
+			"service_account_name": {
+				Type:        framework.TypeString,
+				Description: "Service account name",
+			},
+			"password": {
+				Type:        framework.TypeString,
+				Description: "Password",
+			},
+		},
+		Renew:  b.renewCheckOut,
+		Revoke: b.endCheckOut,
+	}
+}
+
+func (b *backend) renewCheckOut(ctx context.Context, req *logical.Request, _ *framework.FieldData) (*logical.Response, error) {
+	setName := req.Secret.InternalData["set_name"].(string)
+	lock := locksutil.LockForKey(b.checkOutLocks, setName)
+	lock.RLock()
+	defer lock.RUnlock()
+
+	set, err := readSet(ctx, req.Storage, setName)
+	if err != nil {
+		return nil, err
+	}
+	if set == nil {
+		return logical.ErrorResponse(fmt.Sprintf(`%q doesn't exist`, setName)), nil
+	}
+
+	serviceAccountName := req.Secret.InternalData["service_account_name"].(string)
+	checkOut, err := b.LoadCheckOut(ctx, req.Storage, serviceAccountName)
+	if err != nil {
+		return nil, err
+	}
+	if checkOut.IsAvailable {
+		// It's possible that this renewal could be attempted after a check-in occurred either by this entity or by
+		// another user with access to the "manage check-ins" endpoint that forcibly checked it back in.
+		return logical.ErrorResponse(fmt.Sprintf("%s is already checked in, please call check-out to regain it", serviceAccountName)), nil
+	}
+	resp := &logical.Response{Secret: req.Secret}
+	resp.Secret.TTL = set.TTL
+	resp.Secret.MaxTTL = set.MaxTTL
+	return resp, nil
+}
+
+func (b *backend) endCheckOut(ctx context.Context, req *logical.Request, _ *framework.FieldData) (*logical.Response, error) {
+	setName := req.Secret.InternalData["set_name"].(string)
+	lock := locksutil.LockForKey(b.checkOutLocks, setName)
+	lock.Lock()
+	defer lock.Unlock()
+
+	serviceAccountName := req.Secret.InternalData["service_account_name"].(string)
+	if err := b.CheckIn(ctx, req.Storage, serviceAccountName); err != nil {
+		return nil, err
+	}
+	return nil, nil
+}
+
+func (b *backend) pathSetCheckIn() []*framework.Path {
+	return []*framework.Path{
+		{
+			Pattern: libraryPrefix + framework.GenericNameRegex("name") + "/check-in$",
+			Fields: map[string]*framework.FieldSchema{
+				"name": {
+					Type:        framework.TypeLowerCaseString,
+					Description: "Name of the set.",
+					Required:    true,
+				},
+				"service_account_names": {
```

Comment:
I'm wondering if `service_account_names` still makes sense, since we'll allow library sets to be created for schemas other than AD (e.g., OpenLDAP, RACF). I know that we currently target users using [UserPrincipalName in AD secrets engine](https://github.com/hashicorp/vault-plugin-secrets-ad/blob/main/plugin/util/secrets_client.go#L69). UserPrincipalName is an attribute of an [AD user](https://docs.microsoft.com/en-us/powershell/module/activedirectory/new-aduser?view=windowsserver2022-ps). AD [service accounts](https://docs.microsoft.com/en-us/powershell/module/activedirectory/new-adserviceaccount?view=windowsserver2022-ps) do not have the UserPrincipleName attribute. I know that the definition of user vs. service account is pretty loose in LDAP. I think `usernames` could make sense as an alternative name (esp. since static roles in the LDAP secrets engine have a [`username`](https://www.vaultproject.io/api-docs/secret/openldap#username) field).

Looking for opinions on this.

Reply:
TODO: Alias of usernames and fallback to service_account_names

Reply:
Thinking on this some more -- perhaps "service account" does make sense in the context of check-in/check-out. It feels a little odd to think of checking out a "username", even if under the hood, we're managing password rotation for a user-like object in LDAP.

Reply:
If the concern here is that this conflates the term "service account" used elsewhere, could just the word "accounts" be a viable middle-ground here? Though it sounds pretty generic, but figured I'd throw it out there.

Reply:
That's a good suggestion as well. I'm going to leave this as "service account" at this point.